File: java/Kode/Character.java
import java.util.List;
import java.util.ArrayList;

public class Character {
    private String username;
    private int level;
    private String job;
    private int hp;
    private int mana;
    private Weapon equippedWeapon; // Association to Weapon (Polymorphic)
    private List<Skill> learnedSkills; // Association to Skill

    public Character(String username, int level, String job, int hp, int mana) {
        this.username = username;
        this.level = level;
        this.job = job;
        this.hp = hp;
        this.mana = mana;
        this.learnedSkills = new ArrayList<>();
    }

    public void equipWeapon(Weapon weapon) {
        this.equippedWeapon = weapon;
    }

    public void learnSkill(Skill skill) {
        this.learnedSkills.add(skill);
    }

    public String getUsername() {
        return username;
    }

    public void displayCharacter() {
        System.out.println("  [INFO] Username: " + username + ", Job: " + job + ", Level: " + level);
        System.out.println("  [STATS] HP: " + hp + ", Mana: " + mana);

        System.out.println("\n  --- EQUIPPED WEAPON (Association) ---");
        if (equippedWeapon != null) {
            // Polymorphism: Java akan memanggil displayWeaponDetails dari subclass yang tepat
            equippedWeapon.displayWeaponDetails(false); // fullDetail = false (tanpa material)
        } else {
            System.out.println("  No weapon equipped.");
        }

        System.out.println("\n  --- LEARNED SKILLS (Association) ---");
        if (learnedSkills.isEmpty()) {
            System.out.println("  No skills learned.");
        } else {
            for (Skill skill : learnedSkills) {
                skill.displaySkillDetails();
            }
        }
    }
}

File: cpp/Kode/Character.cpp
#include "Character.h"
#include <iostream>

Character::Character(std::string u, int l, std::string j, int hp, int m) : 
    Username(u), Level(l), Job(j), HP(hp), Mana(m) {}

void Character::equipWeapon(Weapon* w) {
    equippedWeapon = w;
}

void Character::learnSkill(Skill* s) {
    learnedSkills.push_back(s);
}

void Character::displayCharacter() const {
    std::cout << "  [CHARACTER] Profile: " << Username << " | Job: " << Job << std::endl;
    std::cout << "Level: " << Level << " | HP: " << HP << " | Mana: " << Mana << std::endl;
    
    // Tampilkan Senjata (Tanpa Material)
    std::cout << "\n-- EQUIPPED WEAPON --" << std::endl;
    if (equippedWeapon) {
        // Polymorphism: Panggil versi tanpa material
        equippedWeapon->displayWeaponWithoutMaterial(); 
    } else {
        std::cout << "  (No weapon equipped)" << std::endl;
    }
    
    // Tampilkan Skill
    std::cout << "\n-- LEARNED SKILLS (" << learnedSkills.size() << ") --" << std::endl;
    if (!learnedSkills.empty()) {
        for (const auto& skill : learnedSkills) {
            skill->displaySkillDetails();
        }
    } else {
        std::cout << "  (No skills learned)" << std::endl;
    }
}